src/github.com/matrix-org/dendrite/mediaapi/writers/download.go

```
@@ -101,3 +118,152 @@ func (r *downloadRequest) Validate() *util.JSONResponse {
+	return &util.JSONResponse{
+		Code: 500,
+		JSON: jsonerror.InternalServerError(),
+	}
+}
+
+// getMediaMetadata queries the database for media metadata
+func (r *downloadRequest) getMediaMetadata(db *storage.Database) (*types.MediaMetadata, error) {
+	mediaMetadata, err := db.GetMediaMetadata(r.MediaMetadata.MediaID, r.MediaMetadata.Origin)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			r.Logger.WithFields(log.Fields{
+				"Origin":  r.MediaMetadata.Origin,
+				"MediaID": r.MediaMetadata.MediaID,
+			}).Info("Media not found in database.")
+			return nil, errDBNotFound
+		}
+		r.Logger.WithError(err).WithFields(log.Fields{
+			"Origin":  r.MediaMetadata.Origin,
+			"MediaID": r.MediaMetadata.MediaID,
+		}).Error("Error querying database for media.")
+		return nil, errDBQuery
+	}
+	return mediaMetadata, nil
+}
+
+// respondFromLocalFile reads a file from local storage and writes it to the http.ResponseWriter
+// Returns a util.JSONResponse error in case of error
+func (r *downloadRequest) respondFromLocalFile(w http.ResponseWriter, absBasePath types.Path) *util.JSONResponse {
+	filePath, err := fileutils.GetPathFromBase64Hash(r.MediaMetadata.Base64Hash, absBasePath)
+	if err != nil {
+		// FIXME: Remove erroneous file from database?
+		r.Logger.WithError(err).Warn("Failed to get file path from metadata")
+		return &util.JSONResponse{
+			Code: 404,
+			JSON: jsonerror.NotFound(fmt.Sprintf("File with media ID %q does not exist", r.MediaMetadata.MediaID)),
+		}
+	}
+	file, err := os.Open(filePath)
+	defer file.Close()
+	if err != nil {
+		// FIXME: Remove erroneous file from database?
+		r.Logger.WithError(err).Warn("Failed to open file")
+		return &util.JSONResponse{
+			Code: 404,
+			JSON: jsonerror.NotFound(fmt.Sprintf("File with media ID %q does not exist", r.MediaMetadata.MediaID)),
+		}
+	}
+	stat, err := file.Stat()
+	if err != nil {
+		// FIXME: Remove erroneous file from database?
+		r.Logger.WithError(err).Warn("Failed to stat file")
```

This code should only be hit if something has gone wrong with the filesystem. I'd probably just return a 500 error in this case.